core/src/main/java/org/apache/calcite/sql/SqlUtil.java

```diff
@@ -1235,6 +1237,65 @@ private static SqlNode createBalancedCall(SqlOperator op, SqlParserPos pos,
     return op.createCall(pos, leftNode, rightNode);
   }
 
+  /**
+   * Creates a table sample spec from the parser information.
+   *
+   * NOTE(jsternberg): This function is really misplaced. I suspect that
+   * this logic doesn't belong in the parser at all and instead should be
+   * in the SQL validator, but that's too large of a refactor to move that
+   * code from the parser to the validator without properly understanding
+   * the consequences for error handling. It might be worth bringing it up
+   * with upstream if/when we try to get our code more compatible with theirs,
+   * but it's not really worth doing that refactor unless we're planning
+   * to contribute the changes back to upstream.
+   *
+   * The parser references the above note. If this code changes, remember to
+   * update the parser comments too.
+   */
+  public static @Nullable SqlSampleSpec createTableSample(@NonNull SqlParserPos pos,
+      boolean isBernoulli, @NonNull SqlNumericLiteral sampleRate, boolean sampleByRows,
+      boolean isRepeatable, int repeatableSeed) {
+    BigDecimal rate = sampleRate.bigDecimalValue();
+    if (rate == null) {
+      return null;
+    }
```

Comment:
Does bigDecimalValue() returning null ever occur or are you just being defensive here?

Reply:
The method itself is annotated that it can return null so this was throwing some warnings so I just decided to be defensive.